File: Roadmap/05 - VALOR Y REFERENCIA/swift/didacdev.swift
import Foundation

// Paso por valor
var valorUno: Int = 3
var valorDos: Int = 5

func variablesValor(variableUno: Int, variableDos: Int) ->  (vairableUno: Int, variableDos: Int){
    let copyVariableUno = variableUno
    let copyVariableDos = variableDos

    return (copyVariableDos, copyVariableUno)
}

let variablesValor: (Int, Int) = variablesValor(variableUno: valorUno, variableDos: valorDos)

let newValorUno = variablesValor.0
let newValorDos = variablesValor.1

print("Variables por valor originales: ")
print("\(valorUno) \(valorDos)")
print("Variables por valor nuevas: ")
print("\(newValorUno) \(newValorDos)")

// Paso por referencia
var referenciaUno: Int = 2
var referenciaDos: Int = 4

func variablesReferencia(variableUno: inout Int, variableDos: inout Int) -> (variableUno: Int, variableDos: Int) {
    let copyVariableUno = variableUno
    let copyVariableDos = variableDos

    variableUno = copyVariableDos
    variableDos = copyVariableUno

    return (variableUno, variableDos)
    
}

let variablesReferencia: (Int, Int) = variablesReferencia(variableUno: &referenciaUno, variableDos: &referenciaDos)

let newReferenciaUno = variablesReferencia.0
let newReferenciaDos = variablesReferencia.1

print("Variables por referencia originales: ")
print("\(referenciaUno) \(referenciaDos)")
print("Variables por valor nuevas: ")
print("\(newReferenciaUno) \(newReferenciaDos)")

File: Roadmap/05 - VALOR Y REFERENCIA/kotlin/didacdev.kt
fun main() {

    // Paso por valor
    var valorUno = 3
    var valorDos = 5

    val variablesValor = variablesValor(valorUno, valorDos)

    val newValorUno = variablesValor.first
    val newValorDos = variablesValor.second

    println("Variables por valor originales: ")
    println("$valorUno $valorDos")
    println("Variables por valor nuevas: ")
    println("$newValorUno $newValorDos")

    // Paso por referencia
    var referenciaUno = arrayOf(2)
    var referenciaDos = arrayOf(4)

    val variablesReferencia = variablesReferencia(referenciaUno, referenciaDos)

    val newReferenciaUno = variablesReferencia.first
    val newReferenciaDos = variablesReferencia.second

    println("Variables por referencia originales: ")
    println("${referenciaUno[0]} ${referenciaDos[0]}")
    println("Variables por referencia nuevas: ")
    println("${newReferenciaUno[0]} ${newReferenciaDos[0]}")
}

fun variablesValor(variableUno: Int, variableDos: Int): Pair<Int, Int> {
    val copyVariableUno = variableUno
    val copyVariableDos = variableDos
    
    return Pair(copyVariableDos, copyVariableUno)
}

fun variablesReferencia(variableUno: Array<Int>, variableDos: Array<Int>): Pair<Array<Int>, Array<Int>> {
    val temp = variableUno[0]
    variableUno[0] = variableDos[0]
    variableDos[0] = temp

    return Pair(variableUno, variableDos)
}